File: packages/face-detection/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/facedetection/FaceDetection.java
package io.capawesome.capacitorjs.plugins.mlkit.facedetection;

import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.face.FaceDetector;
import com.google.mlkit.vision.face.FaceDetectorOptions;
import io.capawesome.capacitorjs.plugins.mlkit.facedetection.classes.ProcessImageOptions;
import io.capawesome.capacitorjs.plugins.mlkit.facedetection.classes.ProcessImageResult;

public class FaceDetection {

    @NonNull
    private final FaceDetectionPlugin plugin;

    public FaceDetection(@NonNull FaceDetectionPlugin plugin) {
        this.plugin = plugin;
    }

    @Nullable
    public InputImage createInputImageFromFilePath(@NonNull String path) {
        try {
            return InputImage.fromFilePath(this.plugin.getContext(), Uri.parse(path));
        } catch (Exception exception) {
            return null;
        }
    }

    public void processImage(ProcessImageOptions options, ProcessImageResultCallback callback) {
        InputImage inputImage = options.getInputImage();
        int performanceMode = options.getPerformanceMode();
        int landmarkMode = options.getLandmarkMode();
        int contourMode = options.getContourMode();
        int classificationMode = options.getClassificationMode();
        float minFaceSize = options.getMinFaceSize();
        boolean enableTracking = options.isTrackingEnabled();

        FaceDetectorOptions.Builder builder = new FaceDetectorOptions.Builder();
        builder.setPerformanceMode(performanceMode);
        builder.setLandmarkMode(landmarkMode);
        builder.setContourMode(contourMode);
        builder.setClassificationMode(classificationMode);
        builder.setMinFaceSize(minFaceSize);
        if (enableTracking) {
            builder.enableTracking();
        }
        FaceDetectorOptions faceDetectorOptions = builder.build();

        final FaceDetector faceDetector = com.google.mlkit.vision.face.FaceDetection.getClient(faceDetectorOptions);
        plugin
            .getActivity()
            .runOnUiThread(() -> {
                faceDetector
                    .process(inputImage)
                    .addOnSuccessListener(faces -> {
                        faceDetector.close();
                        ProcessImageResult result = new ProcessImageResult(faces);
                        callback.success(result);
                    })
                    .addOnCanceledListener(() -> {
                        faceDetector.close();
                        callback.cancel();
                    })
                    .addOnFailureListener(exception -> {
                        faceDetector.close();
                        callback.error(exception);
                    });
            });
    }
}

File: packages/face-detection/ios/Plugin/FaceDetection.swift
import Foundation
import MLKitVision
import MLKitFaceDetection

@objc public class FaceDetection: NSObject {
    public let plugin: FaceDetectionPlugin

    init(plugin: FaceDetectionPlugin) {
        self.plugin = plugin
    }

    @objc func createVisionImageFromFilePath(_ path: String) -> VisionImage? {
        guard let url = URL.init(string: path) else {
            return nil
        }
        if FileManager.default.fileExists(atPath: url.path) {
            guard let image = UIImage.init(contentsOfFile: url.path) else {
                return nil
            }
            return VisionImage.init(
                image: image
            )
        } else {
            return nil
        }
    }

    @objc func processImage(_ options: ProcessImageOptions, completion: @escaping (ProcessImageResult?, Error?) -> Void) {
        let visionImage = options.getVisionImage()
        let performanceMode = options.getPerformanceMode()
        let landmarkMode = options.getLandmarkMode()
        let contourMode = options.getContourMode()
        let classificationMode = options.getClassificationMode()
        let minFaceSize = options.getMinFaceSize()
        let enableTracking = options.isTrackingEnabled()

        let faceDetectorOptions: FaceDetectorOptions = FaceDetectorOptions()
        faceDetectorOptions.performanceMode = FaceDetectorPerformanceMode(rawValue: performanceMode)
        faceDetectorOptions.landmarkMode = FaceDetectorLandmarkMode(rawValue: landmarkMode)
        faceDetectorOptions.contourMode = FaceDetectorContourMode(rawValue: contourMode)
        faceDetectorOptions.classificationMode = FaceDetectorClassificationMode(rawValue: classificationMode)
        faceDetectorOptions.minFaceSize = CGFloat(minFaceSize)
        faceDetectorOptions.isTrackingEnabled = enableTracking

        let faceDetector = FaceDetector.faceDetector(
            options: faceDetectorOptions
        )

        do {
            let faces = try faceDetector.results(
                in: visionImage
            )
            let result = ProcessImageResult(faces: faces)
            completion(result, nil)
        } catch let error {
            completion(nil, error)
        }
    }
}
